pkg/dataobj/internal/dataset/value_encoding_plain.go

```diff
@@ -240,26 +241,56 @@ func (dec *plainBytesDecoder) Decode(s []Value) (int, error) {
 		} else if err != nil {
 			return i, err
 		}
-		s[i] = v
 	}
 	return len(s), nil
 }
 
 // decode decodes a string.
-func (dec *plainBytesDecoder) decode() (Value, error) {
+func (dec *plainBytesDecoder) decode(v *Value) error {
 	sz, err := binary.ReadUvarint(dec.r)
 	if err != nil {
-		return ByteArrayValue([]byte{}), err
+		return err
 	}
 
-	dst := make([]byte, int(sz))
+	dst := getDestinationBuffer(v, int(sz))
 	if _, err := io.ReadFull(dec.r, dst); err != nil {
-		return ByteArrayValue([]byte{}), err
+		return err
 	}
-	return ByteArrayValue(dst), nil
+
+	v.num = sz
+	v.any = (bytearray)(unsafe.SliceData(dst))
+	return nil
 }
 
 // Reset implements [valueDecoder]. It resets the decoder to read from r.
 func (dec *plainBytesDecoder) Reset(r streamio.Reader) {
 	dec.r = r
 }
+
+func getDestinationBuffer(v *Value, sz int) []byte {
+	if v.cap == 0 {
```

Comment:
same concern as above: this should be a method on `Value` to avoid relying on its internal state in other parts of the code